src/main/java/org/truffleruby/core/hash/library/CompactHashStore.java

```diff
@@ -0,0 +1,416 @@
+  (archived at https://archive.ph/wip/sucRY)
+ */
+public class CompactHashStore {
+    //Big Picture :
+    //  (1) A Compact Hash's big idea is storing all key-values in a contiguous array (the KV array) in insertion order
+    //      This enables us to do sequential iteration in insertion order, which is what we want
+    //      But it doesn't allow constant-time lookups
+    //  -----------------------------------------------
+    //  (2) Thus, a Compact Hash also stores the position of every key in the KV array in another array called the index
+    //      That position is stored at a location that depends on the respective key's hash
+    //      Thus given a key we can always quickly find where it's (and we also store the hash itself in the index)
+    //  -----------------------------------------------
+    //  (3) Therefore, Compact Hashes give you both insertion-order iteration as well as the usual const-time guarantees
+    //  ---------------------------------------------------------------------------------------------------------------
+
+
+    private Object[] kvStore;
+    private int kvStoreInsertionPos;
+    /* - We view the index array as consisting of "Slots", each slot is 2 array positions
+      - The first holds the hash of a key and the second holds an offseted index into the KV store (index + 1) where the
+        key is stored
+      - Slots always begin at even positions : index[0] and index[1] is one slot, index[2] and index[3] is another,...*/
+    private int[] index;
+
+    /*
+    Each slot in the index can be in one of 3 states, depending on the value of its second (offset) field
+        - State 1 : Filled, the data in the hash field and the index field is valid,
+                    in this case the index field is >= 1
+        - State 2 : Empty, the data in the hash field and the index field is NOT valid,
+                    and the slot was never filled with valid data,
+                    in this case the index field is INDEX_SLOT_EMPTY_MARKER
+        - State 3 : Deleted, the data in the hash field and the index field is NOT valid,
+                    but the slot used to be occupied with valid data before
+                    in this case the index field is INDEX_SLOT_DELETED_MARKER
+    */
+    private static final int INDEX_SLOT_EMPTY_MARKER = 0;
+    //
+    private static final int INDEX_SLOT_DELETED_MARKER = -1;
+
+
+    public static final int DEFAULT_INITIAL_CAPACITY = 8 ;
+    public static final float THRESHOLD_LOAD_FACTOR_FOR_INDEX_REBUILD = 0.7f;
+
+    private static boolean LOG_FIRST_ACCESS = true;
+    public CompactHashStore(int initialCapacity) {
+        if (LOG_FIRST_ACCESS) {
+            System.out.println("~~~~~~~~~~~~~~~~~~~~~~~ Using Compact Hashes from Hash Tables That Don't Hate You.");
+            LOG_FIRST_ACCESS = false;
+        }
+        //All 0s by default, so all slots marked empty for free
+        this.index = new int[2*initialCapacity];
```

Comment:
This means if there are 8 entries all elements of the index array will be used.
That's probably too tight, i.e., it will cause a lot of collisions in the index array.
So the index array should be bigger most likely.
I looked at `org.graalvm.collections.EconomicMapImpl#createHash` for instance

Reply:
I looked at EconomicMapImpl and it just multiplies by 2, I think would be really wasteful most of the time. 

How about we just overallocate by 1.4 or something, i.e. we allocate 2 * ((int) (initialCapacity * 1.2)) entries for the index array. I feel that's enough for some breathing room without being too wasteful. Or maybe we can overallocate additevely by some constant amount, for example initialCapacity + 64.

Reply:
We need a multiple of two size to be able to `&` instead of `%`.
Let's just use the power of 2 > initialCapacity, and some extra room so I would follow what `EconomicMapImpl#createHash` does:
```java
    private void createHash() {
        int entryCount = size();

        // Calculate smallest 2^n that is greater number of entries.
        int size = getHashThreshold();
        while (size <= entryCount) {
            size <<= 1;
        }

        // Give extra size to avoid collisions.
        size <<= 1;
```

Reply:
That's done in the new diff, great